src/main/webapp/resources/js/pages/projects/samples-metadata-import/redux/importReducer.ts

```diff
@@ -93,77 +94,145 @@ export const saveMetadata = createAsyncThunk<
-                  metadataSaveDetails[index] = {
-                    saved: false,
-                    error: error.response.data.error,
-                  };
-                })
-            );
-          } else {
-            promises.push(
-              createSample({
-                projectId,
-                body: {
-                  name,
-                  metadata: metadataFields,
-                },
-              })
-                .then(() => {
-                  metadataSaveDetails[index] = { saved: true };
-                })
-                .catch((error) => {
-                  metadataSaveDetails[index] = {
-                    saved: false,
-                    error: error.response.data.error,
-                  };
-                })
-            );
-          }
-        }
-      }
-      await Promise.all(promises).then(() => {
-        dispatch(
-          setMetadataSaveDetails(Object.assign({}, metadataSaveDetails))
+    const selectedSampleList = metadata.filter((metadataItem) => {
+      const name: string = metadataItem[sampleNameColumn];
+      return (
+        selectedMetadataKeys.includes(metadataItem.rowKey) &&
+        metadataSaveDetails[name]?.saved !== true
+      );
+    });
+
+    const chunkSize = 10;
+
+    //update existing project samples
+    const updateSamplesPromises: Promise<void>[] = [];
+    const updateSampleList = selectedSampleList
+      .filter((metadataItem) => {
+        const name = metadataItem[sampleNameColumn];
+        const sampleId = metadataValidateDetails[name].foundSampleId;
+        return sampleId;
+      })
+      .map((metadataItem) => {
+        const name = metadataItem[sampleNameColumn];
+        const sampleId = metadataValidateDetails[name].foundSampleId;
+        const metadataFields = Object.entries(metadataItem)
+          .filter(
+            ([key]) =>
+              headers.map((header) => header.name).includes(key) &&
+              key !== sampleNameColumn
+          )
+          .map(([key, value]) => ({
+            field: key,
+            value,
+            restriction: headers.filter((header) => header.name === key)[0]
+              .targetRestriction,
+          }));
+        return { name, sampleId, metadata: metadataFields };
+      });
+    //create a request with a list of project samples to be updated
+    if (updateSampleList.length > 0) {
+      const chunkedUpdateSampleList = chunkArray(updateSampleList, chunkSize);
+      for (const chunk of chunkedUpdateSampleList) {
+        updateSamplesPromises.push(
+          updateSamples({
+            projectId,
+            body: chunk,
+          })
+            .then(() => {
+              chunk
+                .map((item) => item.name)
+                .forEach(
+                  (name) => (metadataSaveDetails[name] = { saved: true })
+                );
+            })
+            .catch((error) => {
+              const { errors } = error.response.data;
```

Comment:
If you have a partial create failure, you will not be setting the saved samples to true. I think changing the server side request to return a status response that lists the outcome for each sample would be the best. In the case of a partial success you could return status code 207. https://learn.microsoft.com/en-us/rest/api/searchservice/http-status-codes

Reply:
My understanding of a multi-status return code is the response includes information on a variety of resources. I would have to go through each sample to determine if it was successful or failed. If a we have a partial failure, we would set the failed sample to not saved and the successful samples to saved. If the user wants to try importing again, it would skip the saved samples and make the request go faster, no? Please see b357be2f267d57cc42ccb5642f9b510c1ff039ae.

Reply:
Yes the multi-status code would only be returned if at least one of the samples to be updated or created were successful. If all samples failed create/update then it would not use the multi status code. Within the response it would be good to note which samples were successfull and which ones failed vs just returned which ones failed (I am just thinking back to null pointer exception bug that I recently fixed that caused a partial response, having a full response means we are confident that a subset of the samples actually saved).

Reply:
Ok, will do.